### DISCUSSION TITLE
How to fetch all details linking to multiple models in dynamoose

### DISCUSSION THREAD

**Discussion by ghost** [dynamoose/dynamoose#1105] [Q&A] [answered]
Hello, I have a confusion in **fetching all users with their respective project details and department details**.
In first API I was able to get one user's detail by passing userId, I used async.series for that and query( because i had a userId which is a partition key).
But in this  second api, can an anyone tell me how can I do for fetching all user details with their respective details? Here I am not passing any params, its just /getAllDetails url in postman

_This is the user schema, it has a projectId_
```
var userSchema = new dynamoose.Schema({
    "userId" :{
        "type" : String,
        "hashKey" : true
        },    
    "projectId" : {
        "type" : String,
        "rangeKey" : true
    }  
});
```
_This is my project schema, it has a departmentId_
```
var projectSchema = new dynamoose.Schema({
    "projectId":{
        "type": String,
        "hashKey" :true
    },
    "departmentId" : {
        "type" : String,
        "rangeKey" :true
    }    
})
```
_And my department Schema_
```
var departmentSchema = new dynamoose.Schema({
    "departmentId" :{
        "type": String,
        "hashKey" : true
    },
    "departmentName": {
        "type" : String,
        "required": true
    }
});
```
_Edited:_
Here is my controller.js
```
            User.scan()
                .exec(async (error, results) => {
                    if (error) {
                        console.error(error);
                    } else {  
                        for(let item=0; item<results.length; item++){
                            projectId = results[item].projectId
                            console.log("1==>"+projectId)
                            projectDetails= await getProjectDetails(projectId)
                            console.log("3==>"+projectDetails);
                           // getDepartmentDetails = await getDepartmentDetails();   
                        }         
                    }
                });

        function getProjectDetails(projectId) {
            Projects
            .query("projectId")
            .eq(projectId)
            .exec((error, results) => {
                if (error) {
                    console.error(error);
                } else {
                    console.log("2==>"+results)
                    return results
                }
            });
        }
```
The new problem here is in the for loop now, await does not wait for the return result from the getProjectDetails(), So it goes like
1 - (projectId)
3 - undefined
2 - (object)
What is the alternate solution for this?

**Comment by fishcharlie** [comment, marked as answer]
@ssagarGurav This is WAY too much information for this question, and this question isn't even related to Dynamoose. The problem you are having is a JavaScript problem.
 
You have `projectDetails= await getProjectDetails(projectId)`. But `getProjectDetails` isn't actually returning anything. You are trying to mix callbacks and promises, which just creates confusion and makes it more complicated.

Here are a few options to fix this:

```
function getProjectDetails(projectId) {
            return Projects
            .query("projectId")
            .eq(projectId)
            .exec();
        }
```

OR

```
async function getProjectDetails(projectId) {
            const results = await Projects
            .query("projectId")
            .eq(projectId)
            .exec();

            return results;
        }
```

OR you could change `getProjectDetails` to accept a callback as well, and call the callback inside of your exec function.

TLDR. don't mix callbacks and promises.

**Reply by ghost** [reply]
Yes, its working properly now! Thanks for the information. I need to brush up more about callbacks and promises. Thanks buddy
